src/config_mem.c

```diff
@@ -0,0 +1,223 @@
+/*
+ * Copyright (C) the libgit2 contributors. All rights reserved.
+ *
+ * This file is part of libgit2, distributed under the GNU GPL v2 with
+ * a Linking Exception. For full terms see the included COPYING file.
+ */
+
+#include "config.h"
+
+#include "config_backend.h"
+#include "config_parse.h"
+#include "config_entries.h"
+
+typedef struct {
+	git_config_backend parent;
+	git_config_entries *entries;
+	git_buf cfg;
+} config_memory_backend;
+
+typedef struct {
+	git_config_entries *entries;
+	git_config_level_t level;
+} config_memory_parse_data;
+
+static int config_error_readonly(void)
+{
+	giterr_set(GITERR_CONFIG, "this backend is read-only");
+	return -1;
```

Comment:
Still longing for a `GIT_ENOTSUP`…

Reply:
Hmm, interesting.  I'm game for an `ENOTSUPPORTED` or whatever, but I worry that it would get really weird really fast, at least here for config.  Since configuration is so low-level, I worry that it would leak out very strangely.  For example, if I called `git_merge` and got an `GIT_ENOTSUPPORTED` back then I would very much not understand what it was that wasn't supported.

I think?  Is this actually the case that you had in mind that you do in fact want?

Reply:
Well it could be mostly given as higher-level layers that the current thing they're trying to do can't work, that would allow that higher-level to preserve the semantics of what happened.

For example, we don't do octopus merges, that can be used to signal that to the user (instead of the generic error we're currently returning).
In the config case, the higher level could try other backends and either swallow the error if another one was able to handle the request or error (as currently done), or return `ENOTSUPPORTED` to be more specific about the failure (in case there's only one in-memory backend loaded, which the user is ultimately responsible for). Which makes it pretty close to the `ENOTFOUND` semantics some codepaths have.

This is pretty minor though, it's just that there are some places where it would maybe be useful to report to the user that they're facing a "limitation" of the API, not a hard error.

Reply:
I agree that `ENOTSUPPORTED` would be a nice to have in some locations, including this one. The PR is already big enough as is right now, if you ask me, so I'd love to postpone this to another PR. Especially so as it might require reviewing of callers, as noted by @ethomson

Reply:
It was just wishful thinking 😉. Since it seems we find that a worthwhile error code, I'll try to create a PR for it.

Reply:
I'm skeptical of run-time feature detection like this. Do we have something that isn't detectable at compile time and isn't just a missing feature or a bug?